Source/WebCore/platform/gstreamer/WebKitFliteSourceGStreamer.cpp

```diff
@@ -0,0 +1,263 @@
+    GST_PAD_SRC,
+    GST_PAD_ALWAYS,
+    GST_STATIC_CAPS("audio/x-raw, "
+        "format = (string) " GST_AUDIO_NE(S16) ", "
+        "layout = (string) interleaved, "
+        "rate = (int) 48000, " "channels = (int) [1, 8]")
+);
+
+#define webkit_flite_src_parent_class parent_class
+WEBKIT_DEFINE_TYPE_WITH_CODE(WebKitFliteSrc, webkit_flite_src, GST_TYPE_BASE_SRC,
+    GST_DEBUG_CATEGORY_INIT(webkit_flite_src_debug, "webkitflitesrc", 0, "flitesrc element"));
+
+static cst_voice* defaultVoice()
+{
+    // The first voice on the registered voice list is the default.
+    return flite_voice_list ? val_voice(val_car(flite_voice_list)) : nullptr;
+}
+
+static void webkitFliteSrcReset(WebKitFliteSrc* src)
+{
+    WebKitFliteSrcPrivate* priv = src->priv;
+
+    DataMutexLocker members { priv->dataMutex };
+    gst_adapter_clear(members->adapter.get());
+    members->didLoadUtterance = false;
+}
+
+static void webkitFliteSrcConstructed(GObject* object)
+{
+    WebKitFliteSrc* src = WEBKIT_FLITE_SRC(object);
+    WebKitFliteSrcPrivate* priv = src->priv;
+
+    /* we operate in time */
+    gst_base_src_set_format(GST_BASE_SRC(src), GST_FORMAT_TIME);
+    gst_base_src_set_blocksize(GST_BASE_SRC(src), -1);
+    gst_base_src_set_automatic_eos(GST_BASE_SRC(src), FALSE);
+
+    DataMutexLocker members { priv->dataMutex };
+    members->adapter = adoptGRef(gst_adapter_new());
+
+    ensureFliteVoicesInitialized();
+}
+
+static GstStateChangeReturn webkitFliteSrcChangeState(GstElement* element, GstStateChange transition)
+{
+    WebKitFliteSrc* src = WEBKIT_FLITE_SRC(element);
+
+    GST_DEBUG_OBJECT(src, "Handling %s transition", gst_state_change_get_name(transition));
+
+    GstStateChangeReturn result = GST_CALL_PARENT_WITH_DEFAULT(GST_ELEMENT_CLASS, change_state, (element, transition), GST_STATE_CHANGE_FAILURE);
+
+    return result;
+}
+
+static gboolean webkitFliteSrcStart(GstBaseSrc* baseSource)
+{
+    WebKitFliteSrc* src = WEBKIT_FLITE_SRC(baseSource);
+    WebKitFliteSrcPrivate* priv = src->priv;
+
+    if (priv->text.isEmpty() || !priv->voice) {
+        GST_ERROR_OBJECT(src, "No utterance provided.");
+        return FALSE;
+    }
+    return TRUE;
+}
+
+static gboolean webkitFliteSrcStop(GstBaseSrc* baseSource)
+{
+    WebKitFliteSrc* src = WEBKIT_FLITE_SRC(baseSource);
+    webkitFliteSrcReset(src);
+    return TRUE;
+}
+
+static GstFlowReturn webkitFliteSrcCreate(GstBaseSrc* baseSource, guint64 offset, guint length, GstBuffer** buffer)
+{
+    UNUSED_PARAM(offset);
+    UNUSED_PARAM(length);
+
+    WebKitFliteSrc* src = WEBKIT_FLITE_SRC(baseSource);
+    WebKitFliteSrcPrivate* priv = src->priv;
+
+    gsize bytes = DEFAULT_SAMPLES_PER_BUFFER * sizeof(gint16) * priv->info.channels;
+    DataMutexLocker members { priv->dataMutex };
+    if (!members->didLoadUtterance) {
+        members->didLoadUtterance = true;
+
+        cst_wave* wave = flite_text_to_wave(priv->text.utf8().data(), priv->voice);
```

Comment:
When is `wave` freed?

Reply:
Deciding when we free cst_wave is tricky. Even before a WebKitFliteSrc instance is created, PlatformSpeechSynthesizer can request available voice lists. I made the cst_wave list static and applied a smart pointer to its elements.